File: app/frontend/src/services/authService.js
// src/services/authService.js

import axios from 'axios';

// API base URL
const API_URL = 'http://localhost:8000/api';

// Token storage keys
const ACCESS_TOKEN_KEY = 'fithub_access_token';
const REFRESH_TOKEN_KEY = 'fithub_refresh_token';
const USER_KEY = 'fithub_user';

// Create axios instance with base URL
const apiClient = axios.create({
  baseURL: API_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Add token to requests if available
apiClient.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem(ACCESS_TOKEN_KEY);
    if (token) {
      config.headers['Authorization'] = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

/**
 * Register a new user
 * @param {Object} userData - User registration data
 * @returns {Promise<Object>} Registered user
 */
export const registerUser = async (userData) => {
  try {
    // Transform data to match API expectations
    const apiData = {
      username: userData.username,
      email: userData.email,
      password: userData.password,
      usertype: userData.userType === 'dietitian' ? 'dietitian' : 'user'
    };

    // Add dietitian data if applicable
    if (userData.userType === 'dietitian' && userData.certificationUrl) {
      apiData.dietitian = {
        certification_url: userData.certificationUrl
      };
    }

    const response = await apiClient.post('/register/', apiData);
    
    // The API doesn't return the user immediately since email verification is required
    // We'll just return the submitted data minus the password
    const { password, ...registeredUser } = userData;
    
    return {
      ...registeredUser,
      message: response.data?.detail || 'Registration successful! Please check your email to verify your account.'
    };
  } catch (error) {
    console.error('Registration error:', error.response?.data || error.message);
    throw new Error(
      error.response?.data?.detail || 
      Object.values(error.response?.data || {})[0]?.[0] || 
      'Registration failed. Please try again.'
    );
  }
};

/**
 * Login a user
 * @param {Object} credentials - Login credentials
 * @returns {Promise<Object>} Logged in user
 */
export const loginUser = async (credentials) => {
  try {
    // For JWT, we need to send email as username
    const loginData = {
      username: credentials.email, // API expects username but wants email
      password: credentials.password
    };

    const response = await apiClient.post('/token/', loginData);
    const { access, refresh } = response.data;
    
    // Store tokens in localStorage
    localStorage.setItem(ACCESS_TOKEN_KEY, access);
    localStorage.setItem(REFRESH_TOKEN_KEY, refresh);
    
    // Get user details from token
    const user = parseJwt(access);
    
    // Create user object
    const userData = {
      id: user.user_id,
      email: credentials.email,
      userType: user.usertype || 'user'
    };
    
    // Store user in localStorage
    localStorage.setItem(USER_KEY, JSON.stringify(userData));
    
    return userData;
  } catch (error) {
    console.error('Login error:', error.response?.data || error.message);
    throw new Error(
      error.response?.data?.detail || 
      'Invalid credentials. Please try again.'
    );
  }
};

/**
 * Logout the current user
 * @returns {Promise<void>}
 */
export const logoutUser = async () => {
  try {
    // Call logout API
    await apiClient.post('/logout/');
  } catch (error) {
    console.error('Logout error:', error.response?.data || error.message);
  } finally {
    // Clear local storage regardless of API response
    localStorage.removeItem(ACCESS_TOKEN_KEY);
    localStorage.removeItem(REFRESH_TOKEN_KEY);
    localStorage.removeItem(USER_KEY);
  }
};

/**
 * Refresh the access token using the refresh token
 * @returns {Promise<string>} New access token
 */
export const refreshToken = async () => {
  try {
    const refresh = localStorage.getItem(REFRESH_TOKEN_KEY);
    if (!refresh) {
      throw new Error('No refresh token available');
    }
    
    const response = await apiClient.post('/token/refresh/', { refresh });
    const { access } = response.data;
    
    // Store new access token
    localStorage.setItem(ACCESS_TOKEN_KEY, access);
    
    return access;
  } catch (error) {
    console.error('Token refresh error:', error.response?.data || error.message);
    // If refresh token is invalid, log out the user
    logoutUser();
    throw error;
  }
};

/**
 * Parse JWT token to get user information
 * @param {string} token - JWT token
 * @returns {Object} Decoded token payload
 */
const parseJwt = (token) => {
  try {
    const base64Url = token.split('.')[1];
    const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
    const jsonPayload = decodeURIComponent(atob(base64).split('').map((c) => {
      return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
    }).join(''));
    
    return JSON.parse(jsonPayload);
  } catch (error) {
    console.error('Error parsing JWT:', error);
    return {};
  }
};

/**
 * Get the current user
 * @returns {Promise<Object>} Current user
 */
export const getCurrentUser = async () => {
  try {
    // Get user from localStorage
    const userJson = localStorage.getItem(USER_KEY);
    if (!userJson) {
      throw new Error('User not found');
    }
    
    return JSON.parse(userJson);
  } catch (error) {
    console.error('Get current user error:', error);
    throw error;
  }
};

/**
 * Check if the user is authenticated
 * @returns {boolean} Authentication status
 */
export const isAuthenticated = () => {
  return localStorage.getItem(ACCESS_TOKEN_KEY) !== null;
};

/**
 * Request password reset
 * @param {string} email - User email
 * @returns {Promise<Object>} Response data
 */
export const requestPasswordReset = async (email) => {
  try {
    const response = await apiClient.post(
            '/request-password-reset-code/',     // ← exact path in urls.py
            { email }
          );
    return response.data;
  } catch (error) {
    console.error('Password reset request error:', error.response?.data || error.message);
    throw new Error(
      error.response?.data?.detail || 
      error.response?.data?.email?.[0] || 
      'Failed to request password reset. Please try again.'
    );
  }
};

/**
 * Reset password with token
 * @param {string} token - Reset token
 * @param {string} newPassword - New password
 * @returns {Promise<Object>} Response data
 */
export const resetPassword = async (token, newPassword) => {
  try {
    const response = await apiClient.post('/reset-password/', {
      token,
      new_password: newPassword
    });
    return response.data;
  } catch (error) {
    console.error('Password reset error:', error.response?.data || error.message);
    throw new Error(
      error.response?.data?.detail || 
      error.response?.data?.token?.[0] || 
      error.response?.data?.new_password?.[0] || 
      'Failed to reset password. Please try again.'
    );
  }
};

/**
 * Request password reset code (6-digit)
 * @param {string} email - User email
 * @returns {Promise<Object>} Response data
 */
export const requestPasswordResetCode = async (email) => {
  try {
    const response = await apiClient.post('/request-reset-code/', { email });
    return response.data;
  } catch (error) {
    console.error('Password reset code request error:', error.response?.data || error.message);
    throw new Error(
      error.response?.data?.detail || 
      error.response?.data?.email?.[0] || 
      'Failed to request reset code. Please try again.'
    );
  }
};

/**
 * Verify reset code and optionally reset password
 * @param {string} email - User email
 * @param {string} resetCode - 6-digit reset code
 * @param {string} newPassword - New password (optional)
 * @returns {Promise<Object>} Response data
 */
export const verifyResetCode = async (email, resetCode, newPassword = null) => {
  try {
    // Create the request data object
    const requestData = {
      email,
      code: resetCode
    };
    
    // Add new password if provided
    if (newPassword) {
      requestData.new_password = newPassword;
    }
    const response = await apiClient.post(
            '/verify-reset-code/',               // already correct
            requestData
          );
    return response.data;
  } catch (error) {
    console.error('Reset code verification error:', error.response?.data || error.message);
    throw new Error(
      error.response?.data?.detail || 
      error.response?.data?.code?.[0] || 
      'Invalid reset code. Please try again.'
    );
  }
};

/**
 * Verify email
 * @param {string} token - Verification token
 * @returns {Promise<Object>} Response data
 */
export const verifyEmail = async (token) => {
  try {
    const response = await apiClient.get(`/verify-email/${token}/`);
    return response.data;
  } catch (error) {
    console.error('Email verification error:', error.response?.data || error.message);
    throw new Error(
      error.response?.data?.detail || 
      'Failed to verify email. The link may be expired or invalid.'
    );
  }
};

export default {
  registerUser,
  loginUser,
  logoutUser,
  getCurrentUser,
  isAuthenticated,
  refreshToken,
  requestPasswordReset,
  resetPassword,
  requestPasswordResetCode,
  verifyResetCode,
  verifyEmail
};

File: mobile/lib/services/auth_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'storage_service.dart';
import '../models/login_response.dart';

class AuthenticationException implements Exception {
  final String message;
  final int? statusCode;

  AuthenticationException(this.message, {this.statusCode});

  @override
  String toString() => message;
}

class AuthService {
  static const String baseUrl = 'https://fithubmp.xyz:8000';

  Future<LoginResponse> login(String email, String password) async {
    try {
      // First, authenticate user with login endpoint
      final loginResponse = await http.post(
        Uri.parse('$baseUrl/api/login/'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'email': email, 'password': password}),
      );

      switch (loginResponse.statusCode) {
        case 200:
          final loginData = LoginResponse.fromJson(
            jsonDecode(loginResponse.body),
          );
          // Save token and userId
          await StorageService.saveJwtAccessToken(loginData.token);
          await StorageService.saveUserId(loginData.userId.toString());
          return loginData;
        case 400:
          throw AuthenticationException(
            'Invalid email or password.',
            statusCode: 400,
          );
        case 401:
          throw AuthenticationException(
            'Invalid credentials. Please check your email and password.',
            statusCode: 401,
          );
        case 429:
          throw AuthenticationException(
            'Too many failed login attempts. Please try again later.',
            statusCode: 429,
          );
        default:
          throw AuthenticationException(
            'An error occurred. Please try again later.',
            statusCode: loginResponse.statusCode,
          );
      }
    } catch (e) {
      throw Exception(e.toString());
    }
  }

  Future<void> register({
    required String username,
    required String email,
    required String password,
    required String usertype,
    String? certificationUrl,
  }) async {
    try {
      Map<String, dynamic> requestBody = {
        'username': username,
        'email': email,
        'password': password,
        'usertype': usertype,
      };

      if (usertype == 'dietitian' &&
          certificationUrl != null &&
          certificationUrl.isNotEmpty) {
        requestBody['dietitian'] = {'certification_url': certificationUrl};
      }

      final response = await http.post(
        Uri.parse('$baseUrl/api/register/'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(requestBody),
      );

      switch (response.statusCode) {
        case 201:
          // User registered successfully
          return;
        case 400:
          final responseBody = jsonDecode(response.body);
          String errorMessage = 'Registration failed. Please check your input.';
          if (responseBody is Map) {
            // Example: concatenate error messages from fields
            List<String> errors = [];
            responseBody.forEach((key, value) {
              if (value is List && value.isNotEmpty) {
                errors.add('$key: ${value.join(', ')}');
              } else if (value is String) {
                errors.add('$key: $value');
              }
            });
            if (errors.isNotEmpty) {
              errorMessage = errors.join('; ');
            } else if (responseBody.containsKey('detail')) {
              errorMessage = responseBody['detail'];
            }
          }
          throw AuthenticationException(errorMessage, statusCode: 400);
        default:
          throw AuthenticationException(
            'An error occurred during registration. Please try again later.',
            statusCode: response.statusCode,
          );
      }
    } catch (e) {
      if (e is AuthenticationException) {
        rethrow;
      }
      throw AuthenticationException(
        'Network error during registration: ${e.toString()}',
      );
    }
  } // End of register method

  Future<void> requestPasswordResetCode(String email) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/request-password-reset-code/'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'email': email}),
      );

      if (response.statusCode != 200) {
        final error = jsonDecode(response.body);
        throw AuthenticationException(
          error['email']?[0] ?? error['detail'] ?? 'Failed to send reset code',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      if (e is AuthenticationException) {
        rethrow;
      }
      throw AuthenticationException('Network error: ${e.toString()}');
    }
  }

  Future<String> verifyResetCode(String email, String resetCode) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/verify-reset-code/'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'email': email, 'code': resetCode}),
      );

      final data = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return data['token'];
      } else {
        // Handle specific error messages from the API
        if (data.containsKey('non_field_errors')) {
          throw AuthenticationException(
            data['non_field_errors'][0],
            statusCode: response.statusCode,
          );
        } else if (data.containsKey('email')) {
          throw AuthenticationException(
            data['email'][0],
            statusCode: response.statusCode,
          );
        } else if (data.containsKey('code')) {
          throw AuthenticationException(
            data['code'][0],
            statusCode: response.statusCode,
          );
        }
        // Default error message
        throw AuthenticationException(
          data['detail'] ?? 'Invalid or expired code',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      if (e is AuthenticationException) {
        rethrow;
      }
      throw AuthenticationException('Network error: ${e.toString()}');
    }
  }

  Future<void> resetPassword(String token, String newPassword) async {
    if (newPassword.length < 8) {
      throw AuthenticationException('Password must be at least 8 characters.');
    }
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/reset-password/'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'token': token, 'new_password': newPassword}),
      );

      if (response.statusCode != 200) {
        final error = jsonDecode(response.body);
        throw AuthenticationException(
          error['non_field_errors']?[0] ?? 'Failed to reset password',
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      if (e is AuthenticationException) {
        rethrow;
      }
      throw AuthenticationException('Network error: ${e.toString()}');
    }
  }

  Future<void> logout(String token) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/logout/'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Token $token',
        },
      );

      switch (response.statusCode) {
        case 200:
          return; // Successful logout
        case 400:
          throw AuthenticationException('Logout failed: No token found');
        case 401:
          throw AuthenticationException(
            'Logout failed: User not authenticated',
          );
        default:
          throw AuthenticationException(
            'An error occurred during logout. Please try again later.',
            statusCode: response.statusCode,
          );
      }
    } catch (e) {
      if (e is AuthenticationException) {
        rethrow;
      }
      throw AuthenticationException(
        'Network error during logout: ${e.toString()}',
      );
    }
  }

  // New method to get JWT access token
  Future<Map<String, String>> getJwtAccessToken(
    String email,
    String password,
  ) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/token/'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'username': email, 'password': password}),
      );

      if (response.statusCode == 200) {
        final responseBody = jsonDecode(response.body);
        if (responseBody['access'] != null && responseBody['refresh'] != null) {
          return {
            'access': responseBody['access'],
            'refresh': responseBody['refresh'],
          };
        } else {
          throw AuthenticationException(
            'Invalid JWT response format.',
            statusCode: response.statusCode,
          );
        }
      } else {
        String errorMessage = 'Failed to obtain JWT tokens.';
        try {
          final errorBody = jsonDecode(response.body);
          if (errorBody['detail'] != null) {
            errorMessage = errorBody['detail'];
          }
        } catch (_) {
          errorMessage =
              'Failed to obtain JWT tokens (status ${response.statusCode})';
        }
        throw AuthenticationException(
          errorMessage,
          statusCode: response.statusCode,
        );
      }
    } catch (e) {
      if (e is AuthenticationException) {
        rethrow;
      }
      throw AuthenticationException(
        'Network error while obtaining JWT tokens: ${e.toString()}',
      );
    }
  }
}
